offload/liboffload/src/OffloadImpl.cpp

```diff
@@ -245,3 +308,298 @@ ol_impl_result_t olGetDeviceInfoSize_impl(ol_device_handle_t Device,
                                           size_t *PropSizeRet) {
   return olGetDeviceInfoImplDetail(Device, PropName, 0, nullptr, PropSizeRet);
 }
+
+TargetAllocTy convertOlToPluginAllocTy(ol_alloc_type_t Type) {
+  switch (Type) {
+  case OL_ALLOC_TYPE_DEVICE:
+    return TARGET_ALLOC_DEVICE;
+  case OL_ALLOC_TYPE_HOST:
+    return TARGET_ALLOC_HOST;
+  case OL_ALLOC_TYPE_SHARED:
+  default:
+    return TARGET_ALLOC_SHARED;
+  }
+}
+
+ol_impl_result_t olMemAlloc_impl(ol_device_handle_t Device,
+                                 ol_alloc_type_t Type, size_t Size,
+                                 void **AllocationOut) {
+  auto Alloc =
+      Device->Device.dataAlloc(Size, nullptr, convertOlToPluginAllocTy(Type));
+  if (!Alloc) {
+    return {OL_ERRC_OUT_OF_RESOURCES,
+            formatv("Could not create allocation on device {0}", Device).str()};
+  }
+
+  *AllocationOut = *Alloc;
+  return OL_SUCCESS;
+}
+
+ol_impl_result_t olMemFree_impl(ol_device_handle_t Device, ol_alloc_type_t Type,
+                                void *Address) {
+  auto Res = Device->Device.dataDelete(Address, convertOlToPluginAllocTy(Type));
+  if (Res) {
+    return {OL_ERRC_OUT_OF_RESOURCES, "Could not free allocation"};
+  }
+  return OL_SUCCESS;
+}
+
+ol_impl_result_t olCreateQueue_impl(ol_device_handle_t Device,
+                                    ol_queue_handle_t *Queue) {
+  auto CreatedQueue = std::make_unique<ol_queue_handle_t_>();
+  auto Err = Device->Device.initAsyncInfo(&(CreatedQueue->AsyncInfo));
+  if (Err) {
+    return OL_ERRC_OUT_OF_RESOURCES;
+  }
+  // TODO: Check error
+  CreatedQueue->Device = Device;
+  CreatedQueue->RefCount = 1;
+  *Queue = CreatedQueue.release();
+  return OL_SUCCESS;
+}
+
+ol_impl_result_t olRetainQueue_impl(ol_queue_handle_t Queue) {
+  Queue->RefCount++;
+  return OL_SUCCESS;
+}
+
+ol_impl_result_t olReleaseQueue_impl(ol_queue_handle_t Queue) {
+  Queue->RefCount--;
+  return OL_SUCCESS;
+}
+
+ol_impl_result_t olFinishQueue_impl(ol_queue_handle_t Queue) {
+  // Host plugin doesn't have a queue set so it's not safe to call synchronize
+  // on it, but we have nothing to synchronize in that situation anyway.
+  if (Queue->AsyncInfo->Queue) {
+    auto Err = Queue->Device->Device.synchronize(Queue->AsyncInfo);
+    if (Err) {
+      return OL_ERRC_OUT_OF_RESOURCES;
+    }
+  }
+
+  // Recreate the stream resource so the queue can be reused
+  // TODO: Would be easier for the synchronization to (optionally) not release
+  // it to begin with.
+  auto Res = Queue->Device->Device.initAsyncInfo(&Queue->AsyncInfo);
+  if (Res) {
+    return OL_ERRC_OUT_OF_RESOURCES;
+  }
+
+  return OL_SUCCESS;
+}
+
+ol_impl_result_t olWaitEvent_impl(ol_event_handle_t Event) {
+  auto Res = Event->Device->Device.syncEvent(Event->EventInfo);
+  if (Res) {
+    return OL_ERRC_OUT_OF_RESOURCES;
+  }
```

Comment:
These braces, also other places, should not be here.

Reply:
These should all be fixed now